paddle/fluid/framework/details/inplace_op_pass.cc

```diff
@@ -0,0 +1,362 @@
+    }
+  }
+
+  // release node of unused var in graph
+  for (auto* node : var_nodes_[var]) {
+    graph->RemoveNode(node);
+  }
+  var_nodes_.at(var).clear();
+}
+
+void InplacePass::TryInplaceOpInputOutput(ir::Node* op) const {
+  PADDLE_ENFORCE(op->Op() != nullptr && op->Op()->Block() != nullptr,
+                 "op_desc is nullptr");
+  // 3 pre-requirments need to meet if the op want to inplaced.
+  // 1. infer_inplace_ is registered.
+  auto* op_desc = op->Op();
+  auto& infer_inplace =
+      OpInfoMap::Instance().Get(op_desc->Type()).infer_inplace_;
+  if (!static_cast<bool>(infer_inplace)) return;
+  PADDLE_ENFORCE(static_cast<bool>(infer_inplace),
+                 "%s's infer_inplace has not been registered", op_desc->Type());
+
+  auto* block = op_desc->Block();
+  auto in_to_outs = infer_inplace(*op_desc, block);
+
+  auto& all_ops = view_.AllOps();
+  auto cursor = std::find(all_ops.begin(), all_ops.end(), op);
+  size_t idx = std::distance(all_ops.begin(), cursor);
+
+  for (auto& pair : in_to_outs) {
+    auto& in_var_name = pair.first;
+    auto& out_var_name = pair.second;
+    auto* in_node = view_.GetNodeByName(in_var_name, op->inputs);
+    auto* out_node = view_.GetNodeByName(out_var_name, op->outputs);
+    // 2. there is no external pending op on the input node
+    if (view_.PendingOpsOnVar(in_node).size() > 1) {
+      VLOG(3) << string::Sprintf(
+          "!!! %s input has external dependency, can not inplaced,  %s => %s "
+          "skiped",
+          op->Name(), out_var_name, in_var_name);
+      continue;
+    }
+    // 3. if output reuse input inplaced, the dependency group is not changed.
+    // For detail, check
+    // the function description in "OutConnectInputByCtrlVar"
+    if (view_.OutConnectInputByCtrlVar(in_node, out_node)) {
+      VLOG(3) << string::Sprintf(
+          "!!! %s input output connect by ctrl var, cannot inplaced,  %s => %s "
+          "skiped",
+          op->Name(), out_var_name, in_var_name);
+      continue;
+    }
+    VLOG(3) << string::Sprintf("!!! %s,  %s => %s inplaced", op->Name(),
+                               out_var_name, in_var_name);
+    InplaceModifyDesc(out_var_name, in_var_name, idx);
+    InplaceModifyVar(out_var_name, in_var_name, idx, graph_);
+  }
+}
+
+ir::Node* GraphView::GetNodeByName(const std::string& name,
+                                   const std::vector<ir::Node*>& nodes) const {
+  ir::Node* node = nullptr;
+  for (auto* it : nodes) {
+    if (!it->IsVar() || it->IsCtrlVar() || it->Var() == nullptr) continue;
+    if (it->Name() == name) {
+      node = it;
+      break;
+    }
+  }
+  PADDLE_ENFORCE(node != nullptr,
+                 string::Sprintf("Not found var %s in nodes!", name));
+  return node;
+}
+
+std::vector<ir::Node*> GraphView::PendingOpsOnVar(ir::Node* node) {
+  return node->outputs;
+}
+
+void GraphView::Build(ir::Graph* g) { ops_ = SortOpLikeDescOrder(*g); }
+
+const std::vector<ir::Node*> GraphView::AllOps() { return ops_; }
```

Comment:
return reference?

Reply:
done.

Reply:
it has been fixed. 
https://github.com/PaddlePaddle/Paddle/blob/ad61e1b22c1625db7e096d207c4240fdfea9b2b8/paddle/fluid/framework/details/inplace_op_pass.h#L36

const std::vector<ir::Node*>& 

return reference